### ISSUE TITLE
HTTP and FTP staging do not work with any remote executor

### ISSUE THREAD

**Issue by benclifford** [Parsl/parsl#396]
[ec2-c.py.txt](https://github.com/Parsl/parsl/files/2210332/ec2-c.py.txt)
The attached code throws this exception:

```
2018-07-19 14:27:15 parsl.dataflow.dflow:565 [INFO]  Task 1 submitted for App _http_stage_in, waiting on tasks []
INFO:parsl.dataflow.dflow:Task 1 submitted for App _http_stage_in, waiting on tasks []
Traceback (most recent call last):
  File "./ec2-c.py", line 80, in <module>
    future = foo(inputs=[myfile])
  File "/home/benc/parsl/src/parsl/parsl/app/app_factory.py", line 76, in __call__
    return app_obj(*args, **kwargs)
  File "/home/benc/parsl/src/parsl/parsl/app/bash_app.py", line 150, in __call__
    **self.kwargs)
  File "/home/benc/parsl/src/parsl/parsl/dataflow/dflow.py", line 552, in submit
    self._add_input_deps(executor, args, kwargs)
  File "/home/benc/parsl/src/parsl/parsl/dataflow/dflow.py", line 384, in _add_input_deps
    inputs[idx] = f.stage_in(executor)
  File "/home/benc/parsl/src/parsl/parsl/data_provider/files.py", line 93, in stage_in
    return self.dman.stage_in(self, executor)
  File "/home/benc/parsl/src/parsl/parsl/data_provider/data_manager.py", line 132, in stage_in
    app_fut = stage_in_app(working_dir, outputs=[file])
  File "/home/benc/parsl/src/parsl/parsl/app/app_factory.py", line 76, in __call__
    return app_obj(*args, **kwargs)
  File "/home/benc/parsl/src/parsl/parsl/app/python_app.py", line 49, in __call__
    **kwargs)
  File "/home/benc/parsl/src/parsl/parsl/dataflow/dflow.py", line 581, in submit
    task_id, func, *new_args, **kwargs)
  File "/home/benc/parsl/src/parsl/parsl/dataflow/dflow.py", line 353, in launch_task
    exec_fu = executor.submit(executable, *args, **kwargs)
  File "/home/benc/parsl/src/parsl/parsl/executors/ipp.py", line 228, in submit
    return self.lb_view.apply_async(*args, **kwargs)
  File "/home/benc/parsl/virtualenv/lib/python3.5/site-packages/ipyparallel/client/view.py", line 220, in apply_async
    return self._really_apply(f, args, kwargs, block=False)
  File "<decorator-gen-137>", line 2, in _really_apply
  File "/home/benc/parsl/virtualenv/lib/python3.5/site-packages/ipyparallel/client/view.py", line 52, in sync_results
    ret = f(self, *args, **kwargs)
  File "<decorator-gen-136>", line 2, in _really_apply
  File "/home/benc/parsl/virtualenv/lib/python3.5/site-packages/ipyparallel/client/view.py", line 37, in save_ids
    ret = f(self, *args, **kwargs)
  File "/home/benc/parsl/virtualenv/lib/python3.5/site-packages/ipyparallel/client/view.py", line 1042, in _really_apply
    metadata=metadata)
  File "/home/benc/parsl/virtualenv/lib/python3.5/site-packages/ipyparallel/client/client.py", line 1418, in send_apply_request
    item_threshold=self.session.item_threshold,
  File "/home/benc/parsl/virtualenv/lib/python3.5/site-packages/ipyparallel/serialize/serialize.py", line 186, in pack_apply_message
    msg = serialize_object(f)
  File "/home/benc/parsl/virtualenv/lib/python3.5/site-packages/ipyparallel/serialize/serialize.py", line 125, in serialize_object
    buffers.insert(0, pickle.dumps(cobj, PICKLE_PROTOCOL))
  File "stringsource", line 2, in zmq.backend.cython.socket.Socket.__reduce_cython__
TypeError: no default __reduce__ due to non-trivial __cinit__
```

**Comment by yadudoc** [comment]
I'm able to reproduce this with `parsl.configs.local_ipp`

**Comment by benclifford** [comment]
I had a play with this: `self._http_stage_in` is trying to take `self` across the network with itself. I moved _http_stage_in to a top level declaration, away from the class (as `self` is not needed) and I can get further in execution.

**Comment by benclifford** [comment]
I think this is likely to affect globus staging too, but I haven't dug into it.

**Comment by benclifford** [comment]
This was fixed in #479 and has been tested in CI for http since #877. FTP is not tested in CI because the travis environment does not support that, but will be tested if you run the test suite against parsl/tests/configs/htex_local.py without `-k not cleannet`. I've just verified that latter works from my laptop.
